src/core/common/feature_experiment_flag.hpp

```diff
@@ -0,0 +1,73 @@
+/*
+ *  Copyright (c) 2022, The OpenThread Authors.
+ *  All rights reserved.
+ *
+ *  Redistribution and use in source and binary forms, with or without
+ *  modification, are permitted provided that the following conditions are met:
+ *  1. Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *  2. Redistributions in binary form must reproduce the above copyright
+ *     notice, this list of conditions and the following disclaimer in the
+ *     documentation and/or other materials provided with the distribution.
+ *  3. Neither the name of the copyright holder nor the
+ *     names of its contributors may be used to endorse or promote products
+ *     derived from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ *  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ *  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ *   This file includes feature runtime flag related definitions.
+ */
+
+#ifndef RUNTIME_FLAG_HPP_
+#define RUNTIME_FLAG_HPP_
+
+#include "openthread-core-config.h"
+
+#include "common/log.hpp"
+#include <openthread/feature_experiment_flag.h>
+#include <openthread/platform/feature_experiment_flag.h>
+
+namespace ot {
+
+#define IS_FEATURE_ENABLED_IN_COMPILE(FEATURE) OPENTHREAD_CONFIG_##FEATURE##_ENABLE
+
+/**
+ *  Define the feature flag, and use IS_FEATURE_ENABLED to guard your feature logic.
+ *  We assume that the compile and experiment flag have the same feature name.
+ *  Sample usage:
+ *   // Use IS_FEATURE_ENABLED MACRO to guard new feature logic.
+ *   if (IS_FEATURE_ENABLED(feature)) {
+ *     // Add new feature logic.
+ *   }
+ */
+#if OPENTHREAD_CONFIG_FEATURE_EXPERIMENT_FLAG_ENABLE
+#define IS_FEATURE_ENABLED(FEATURE) FeatureExperimentFlag::IsFeatureEnabled(OT_FEATURE_##FEATURE)
+#else
+#define IS_FEATURE_ENABLED(FEATURE) 1
+#endif
```

Comment:
Thanks @morningboata.  

Is the goal of the PR to add mechanism to enable/disable certain features/behaviors at run-time (for some subset of features)?

If so, I suggest it may be simpler and more straight-forward  to add public APIs to control this:
- Some existing features already provide this, e.g. `BORDER_ROUTING` we have `otBorderRoutingSetEnabled()` to enable the functionality at run-time, similarly for SRP server, and client, ... 
- The example you mention is `NAT64_TRANSLATOR`. 
  -  We can add new API `otNat64TranslationSetEnabled()` for this instead of using `IS_FEATURE_ENABLED()` model.
   - We can still use `if()` checks to see if the behavior is enabled (at run-time) before calling its methods. For example:
```c++
Error Ip6::ProcessReceiveCallback(...) 
{
    // ...

#if OPENTHREAD_CONFIG_NAT64_TRANSLATOR_ENABLE
    if (Get<Nat64::Translator>().IsEnabled())
    { 
        // Use Translator methods only when enabled.
    }
    else
#endif
    {
        // When feature is not enabled (at compile or run-time).
        // ...
    }
}
```

which I think is simpler to read and follow compared to:
```c++
Error Ip6::ProcessReceiveCallback(...) {
...
#if IS_FEATURE_ENABLED_IN_COMPILE(NAT64_TRANSLATOR)
   // Use IS_FEATURE_ENABLED to guard new feature logic.
   if (IS_FEATURE_ENABLED(NAT64_TRANSLATOR)) {
   // Add new feature logic.
   }
#endif
...
}
```

- Another advantage I see is that with enable/disable API, the API will be next to other related APIs of the feature (same file/module) vs. having it be determined in a platform `otPlatIsFeatureEnabled()` API call (shared for all).